library/common/network/configurator.cc

```diff
@@ -60,21 +61,42 @@ namespace {
 #define SUPPORTS_GETIFADDRS
 #endif
 
-std::atomic<envoy_network_t> MobileUtility::preferred_network_{ENVOY_NET_GENERIC};
+SINGLETON_MANAGER_REGISTRATION(network_configurator);
 
-void MobileUtility::setPreferredNetwork(envoy_network_t network) { preferred_network_ = network; }
+constexpr absl::string_view BaseDnsCache = "base_dns_cache";
 
-envoy_network_t MobileUtility::getPreferredNetwork() { return preferred_network_.load(); }
+std::atomic<envoy_network_t> Configurator::preferred_network_{ENVOY_NET_GENERIC};
 
-std::vector<std::string> MobileUtility::enumerateV4Interfaces() {
+envoy_network_t Configurator::setPreferredNetwork(envoy_network_t network) {
+  return preferred_network_.exchange(network);
+}
+
+envoy_network_t Configurator::getPreferredNetwork() { return preferred_network_.load(); }
+
+void Configurator::refreshDns(envoy_network_t network) {
+  // refreshDns is intended to be queued on Envoy's event loop, whereas preferred_network_ is
+  // updated synchronously. In the event that multiple refreshes become queued on the event loop,
+  // this avoids triggering a refresh for a non-current network.
+  // Note this does NOT completely prevent parallel refreshes from being triggered in multiple
+  // flip-flop scenarios.
+  if (network != preferred_network_.load()) {
+    return;
+  }
+  // TODO(goaway): track event here or are there existing signals we can use?
+  if (auto dns_cache = dns_cache_manager_->lookUpCacheByName(BaseDnsCache)) {
```

Comment:
I wonder if it would be good to fire an event in the case:
- the cache is not present
- when we force refresh

Reply:
I was wondering if we were firing an event in the actual refresh implementation (which might be preferable). But if we're not, I can fire it here.

Reply:
It is not. I can add it there. Let's add one for the cache missing though per our other [thread](https://github.com/envoyproxy/envoy-mobile/pull/1816#discussion_r715211592)

Reply:
Adding the force refresh event here: https://github.com/envoyproxy/envoy/pull/18247/files

Reply:
Thanks @junr03